From 1e6e3c36332a970d7158e8dce8b77f901b45babe Mon Sep 17 00:00:00 2001
From: Delphine CC Chiu <Delphine_CC_Chiu@wiwynn.com>
Date: Fri, 22 Sep 2023 09:43:49 +0800
Subject: [PATCH 02/18] PSUSensor: Fix error for decimal part of scalefactor

Customized scale factor declaration from schema in entity-manager,
declared as type "number." If a number is used in PSUSensor, it loses
the decimal part when transforming the value from variant to unsigned
int, resulting in incorrect calculations.

For example, a sensor named "ADC_TEST" is set as below:

{
    "Address": 31,
    "Bus": 31,
    "Labels": [
        "in6"
    ],
    "Name": "FANBOARD_ADC",
    "Type": "ADC128D818",
    "in6_Name": "ADC_TEST",
    "in6_Scale": 31.9458
}

but scale factor in debug print is 31, not 31.9458:

Jul 28 22:50:36 bmc psusensor[788]: DEBUG_FACTOR:31

The more reasonable way to resolve the issue is to use a variant to
float or double, as adcsensor has been doing so far.

Tested:
Get adc128d818 and ina233 reading from PSUSensor by dbus introspect

Change-Id: If3692ab6b0c790072064a44001e9ef58cb50024c
Signed-off-by: Delphine CC Chiu <Delphine_CC_Chiu@wiwynn.com>
---
 src/psu/PSUSensor.cpp     |  2 +-
 src/psu/PSUSensor.hpp     |  6 +++---
 src/psu/PSUSensorMain.cpp | 15 ++++++++++-----
 3 files changed, 14 insertions(+), 9 deletions(-)

diff --git a/src/psu/PSUSensor.cpp b/src/psu/PSUSensor.cpp
index b3ee416..597e889 100644
--- a/src/psu/PSUSensor.cpp
+++ b/src/psu/PSUSensor.cpp
@@ -50,7 +50,7 @@ PSUSensor::PSUSensor(
     boost::asio::io_context& io, const std::string& sensorName,
     std::vector<thresholds::Threshold>&& thresholdsIn,
     const std::string& sensorConfiguration, const PowerState& powerState,
-    const std::string_view sensorUnits, unsigned int factor, double max,
+    const std::string_view sensorUnits, double factor, double max,
     double min, double offset, const std::string& label, size_t tSize,
     double pollRate, const std::shared_ptr<I2CDevice>& i2cDevice) :
     Sensor(escapeName(sensorName), std::move(thresholdsIn), sensorConfiguration,
diff --git a/src/psu/PSUSensor.hpp b/src/psu/PSUSensor.hpp
index 3411833..8a43ef1 100644
--- a/src/psu/PSUSensor.hpp
+++ b/src/psu/PSUSensor.hpp
@@ -28,7 +28,7 @@ class PSUSensor : public Sensor, public std::enable_shared_from_this<PSUSensor>
               std::vector<thresholds::Threshold>&& thresholds,
               const std::string& sensorConfiguration,
               const PowerState& powerState, std::string_view sensorUnits,
-              unsigned int factor, double max, double min, double offset,
+              double factor, double max, double min, double offset,
               const std::string& label, size_t tSize, double pollRate,
               const std::shared_ptr<I2CDevice>& i2cDevice);
     ~PSUSensor() override;
@@ -53,7 +53,7 @@ class PSUSensor : public Sensor, public std::enable_shared_from_this<PSUSensor>
     boost::asio::random_access_file inputDev;
     boost::asio::steady_timer waitTimer;
     std::string path;
-    unsigned int sensorFactor;
+    double sensorFactor;
     double sensorOffset;
     thresholds::ThresholdTimer thresholdTimer;
     void restartRead();
@@ -75,6 +75,6 @@ struct PSUProperty
     std::string_view labelTypeName;
     double maxReading;
     double minReading;
-    unsigned int sensorScaleFactor;
+    double sensorScaleFactor;
     double sensorOffset;
 };
diff --git a/src/psu/PSUSensorMain.cpp b/src/psu/PSUSensorMain.cpp
index bcba555..0db53ec 100644
--- a/src/psu/PSUSensorMain.cpp
+++ b/src/psu/PSUSensorMain.cpp
@@ -832,7 +832,7 @@ static void createSensorsCallback(
                 try
                 {
                     sensorScaleFactor = std::visit(
-                        VariantToUnsignedIntVisitor(), findCustomScale->second);
+                        VariantToDoubleVisitor(), findCustomScale->second);
                 }
                 catch (const std::invalid_argument&)
                 {
@@ -841,7 +841,8 @@ static void createSensorsCallback(
                 }
 
                 // Avoid later division by zero
-                if (sensorScaleFactor > 0)
+                if (sensorScaleFactor != 0.0 &&
+                    std::isfinite(sensorScaleFactor))
                 {
                     customizedScale = true;
                 }
@@ -963,7 +964,7 @@ static void createSensorsCallback(
             // Similarly, if sensor scaling factor is being customized,
             // then the below power-of-10 constraint becomes unnecessary,
             // as config should be able to specify an arbitrary divisor.
-            unsigned int factor = sensorScaleFactor;
+            double factor = sensorScaleFactor;
             if (!customizedScale)
             {
                 // Preserve existing usage of hardcoded labelMatch table below
@@ -980,8 +981,12 @@ static void createSensorsCallback(
                 auto findScaleFactor = baseConfig->find(strScaleFactor);
                 if (findScaleFactor != baseConfig->end())
                 {
-                    factor = std::visit(VariantToIntVisitor(),
-                                        findScaleFactor->second);
+                    double newFactor = std::visit(VariantToDoubleVisitor(),
+                                                  findScaleFactor->second);
+                    if (newFactor != 0.0 && std::isfinite(newFactor))
+                    {
+                        factor = newFactor;
+                    }
                 }
 
                 lg2::debug(
-- 
2.52.0

